### PR TITLE
Update driver/input/ist415

### PR DESCRIPTION
**Pull request by ewoodev**

### CODE REVIEW

File: apps/examples/touchscreen/Makefile [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+# either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+#
+###########################################################################
+############################################################################
+# apps/system/lcd_tune/Makefile
```
Comment (jeongarmy) [inline apps/examples/touchscreen/Makefile:19]: Could you update file path?

File: os/drivers/input/ist415.c [hunk 2]
```diff
@@ -39,246 +39,856 @@
...
+	ist415_disable_touch,	/* disable */
+	ist415_cmd				/* cmd */
+};
 
 /****************************************************************************
  * Private Functions
  ****************************************************************************/
 
+/****************************************************************************
+ * Name: ist415_irq_handler_work
```
Comment (anjana348) [inline os/drivers/input/ist415.c:104]: function name different

File: os/drivers/input/ist415.c [hunk 3]
```diff
@@ -39,246 +39,856 @@
...
-		if (ret != 1) {
-			touchdbg("ERROR: I2C write failed\n");
-			return ret;
-		}
-		ret = i2c_read(i2c, &config, touch_event + EVENT_PACKET_SIZE, touch_point * EVENT_PACKET_SIZE);
-		if (ret != touch_point * EVENT_PACKET_SIZE) {
-			touchdbg("ERROR: I2C read failed\n");
+
+	if ((left_e > 0) && (left_e < 0x1F)) {
+		ist415_i2c_read(dev, HCOM_GET_EVENT_ALL, (uint8_t *)(touch_event + EVENT_PACKET_SIZE), left_e * EVENT_PACKET_SIZE);
```
Comment (anjana348) [inline os/drivers/input/ist415.c:231]: ret = ist415_i2c_read(..

File: os/drivers/input/ist415.c [hunk 4]
```diff
@@ -39,246 +39,856 @@
...
+		if (ret) {
+			touchdbg("Fail to read left data\n");
```
Comment (anjana348) [inline os/drivers/input/ist415.c:233]: free(touch_event) in case of return from function

File: os/drivers/input/ist415.c [hunk 5]
```diff
@@ -39,246 +39,856 @@
...
 			return ret;
 		}
-	} else if (touch_point >= 0x1F) {
-		touchdbg("ERROR:touch events bufer overflow\n");
+	} else if (left_e >= 0x1F) {
+		touchdbg("Touch events buffer overflow\n");
```
Comment (anjana348) [inline os/drivers/input/ist415.c:237]: same as above

File: os/drivers/input/ist415.c [hunk 6]
```diff
@@ -39,246 +39,856 @@
...
+	touchvdbg("%s %s\n", __func__, on ? "on" : "off");
+
+	msleep(50);
+}
+
+/****************************************************************************
+ * Name: ist415_power_mode
+ ****************************************************************************/
+
+void ist415_power_mode(struct ist415_dev_s *dev, uint8_t data)
```
Comment (anjana348) [inline os/drivers/input/ist415.c:816]: static functions if not used elsewhere

File: os/drivers/input/touchscreen.c [hunk 7]
```diff
@@ -232,17 +238,18 @@ static int touch_ioctl(FAR struct file *filep, int cmd, unsigned long arg)
...
 		case TSIOC_SETAPPNOTIFY: {
 			struct touch_set_callback_s *touch_app = (struct touch_set_callback_s *)arg;
 			priv->app_touch_point_buffer = touch_app->touch_points;
 			priv->is_touch_detected = touch_app->is_touch_detected;
 			touchvdbg("App notification callback register is successful\n");
+			priv->ops->touch_enable(priv);
 		}
 		break;
 #endif
+#ifdef CONFIG_TOUCH_CALLBACK
```
Comment (anjana348) [inline os/drivers/input/touchscreen.c:252]: Why is TSIOC_DISABLE and enable inside touch_Callback condition only

Comment (ewoodev) [reply os/drivers/input/touchscreen.c:252]: The apps is already having controllability, the app can don't read touch.
So if CONFIG_TOUCH_CALLBACK is disable, we don't need to provide enable and disable

File: os/drivers/input/ist415.c (hunk 3 above)
Comment (ewoodev) [reply os/drivers/input/ist415.c]: thanks, fixed

File: os/drivers/input/ist415_update.c [hunk 8]
```diff
@@ -0,0 +1,667 @@
...
+int ist415_check_fw(struct ist415_dev_s *dev)
+{
+	struct firmware firmware;
+	struct ist415_fw *fw = &dev->fw;
+	int retry = 0;
+	int ret = 0;
+
+	ret = request_firmware(&firmware, FIRMWARE_PATH_WASHING_MACHINE, &dev);
+	if (ret < 0) {
+		touchdbg("Don't request firmware\n");
```
Comment (anjana348) [inline os/drivers/input/ist415_update.c:542]: firmware->data is not released as release_firmware not called here

File: os/drivers/input/ist415_update.c [hunk 9]
```diff
@@ -0,0 +1,667 @@
...
+	if (argc < 4) {
+		return -EINVAL;
+	}
+
+	path = argv[2];
+	cal = atoi(argv[3]);
+
+	ret = request_firmware(&firmware, path, &dev);
+	if (ret < 0) {
+		touchdbg("do not request firmware\n");
```
Comment (anjana348) [inline os/drivers/input/ist415_update.c:620]: firmware->data is not released as release_firmware not called here

File: os/drivers/input/ist415_misc.c [hunk 10]
```diff
@@ -0,0 +1,658 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+/****************************************************************************
+ * drivers/input/ist415.c
```
Comment (anjana348) [inline os/drivers/input/ist415_misc.c:19]: file name ist415_misc.c

File: os/drivers/input/ist415_update.c (hunk 9 above)
Comment (ewoodev) [reply os/drivers/input/ist415_update.c:620]: Fixed!

File: os/drivers/input/ist415_update.c (hunk 8 above)
Comment (ewoodev) [reply os/drivers/input/ist415_update.c:542]: I added free in error case in request_firmware()

File: os/drivers/input/touchscreen.c [hunk 11]
```diff
@@ -132,7 +135,15 @@ static int touch_open(FAR struct file *filep)
...
 	if (priv->crefs == 0) {
-		priv->ops->touch_enable(priv);
+		if (priv->ops && priv->ops->touch_enable) {
+			priv->ops->touch_enable(priv);
+		}
+
+		priv->tp_buf.head = 0;
+		priv->tp_buf.tail = 0;
+
+		priv->tp_buf.buffer = (struct touch_sample_s *)kmm_zalloc(sizeof(struct touch_sample_s) * CONFIG_TOUCH_BUFSIZE);
```
Comment (anjana348) [inline os/drivers/input/touchscreen.c:143]: allocation failure not checked

File: os/drivers/input/ist415_update.c [hunk 12]
```diff
@@ -0,0 +1,670 @@
...
+	uint8_t sub2 = ISP_CMD_READ;
+	uint32_t addr = 0;
+	int wsize = sizeof(cmd) + sizeof(sub1) + sizeof(sub2) + sizeof(addr);
+	uint8_t *buf;
+	uint32_t buf32 = ISP_STATUS_BUSY;
+	uint32_t status;
+	int ret = 0;
+	int retry = 200;
+
+	buf = (uint8_t *)kmm_malloc(wsize);
```
Comment (anjana348) [inline os/drivers/input/ist415_update.c:99]: malloc failure not checked

File: os/drivers/input/ist415_update.c [hunk 13]
```diff
@@ -0,0 +1,670 @@
...
+{
+	struct i2c_dev_s *i2c = dev->i2c;
+	struct i2c_config_s *config = dev->i2c_config;
+	uint8_t cmd = HCOM_ISP;
+	int hsize = sizeof(cmd) + sizeof(sub1) + sizeof(sub2) + sizeof(addr);
+	int wsize = hsize + (size * sizeof(uint32_t));
+	uint8_t *buf = NULL;
+	int ret = 0;
+
+	buf = (uint8_t *)kmm_malloc(wsize);
```
Comment (anjana348) [inline os/drivers/input/ist415_update.c:146]: malloc allocation failure not checked

File: os/drivers/input/touchscreen.c [hunk 14]
```diff
@@ -132,7 +133,15 @@ static int touch_open(FAR struct file *filep)
 
 	touch_semtake(&priv->sem, false);
 	if (priv->crefs == 0) {
-		priv->ops->touch_enable(priv);
+		if (priv->ops && priv->ops->touch_enable) {
```
Comment (sunghan-chang) [inline os/drivers/input/touchscreen.c:136]: Hmmmm.
If we get the else case, I mean priv->ops is null or priv->ops->touch_enable is null, no need to anything?

File: os/drivers/input/touchscreen.c [hunk 15]
```diff
@@ -132,7 +133,15 @@ static int touch_open(FAR struct file *filep)
...
+			priv->ops->touch_enable(priv);
+		}
+
+		priv->tp_buf.head = 0;
+		priv->tp_buf.tail = 0;
+
+		priv->tp_buf.buffer = (struct touch_sample_s *)kmm_zalloc(sizeof(struct touch_sample_s) * CONFIG_TOUCH_BUFSIZE);
```
Comment (sunghan-chang) [inline os/drivers/input/touchscreen.c:143]: Don't you need to check null for this allocation?

File: os/drivers/input/ist415_misc.c [hunk 16]
```diff
@@ -0,0 +1,764 @@
...
+void ist415_run_intr_debug(struct ist415_dev_s *dev)
+{
+	uint32_t *buf32;
+	int ret = 0;
+
+	if (dev->intr_debug_size <= 0) {
+		return;
+	}
+
+	buf32 = (uint32_t *)kmm_malloc(dev->intr_debug_size);
```
Comment (anjana348) [inline os/drivers/input/ist415_misc.c:382]: malloc failure not checked

File: os/drivers/input/touchscreen.c [hunk 17]
```diff
@@ -123,16 +125,27 @@ static inline void touch_semgive(sem_t *sem)
...
 	touch_semtake(&priv->sem, false);
 	if (priv->crefs == 0) {
-		priv->ops->touch_enable(priv);
+
+		priv->tp_buf.head = 0;
+		priv->tp_buf.tail = 0;
+
+		priv->tp_buf.buffer = (struct touch_sample_s *)kmm_zalloc(sizeof(struct touch_sample_s) * CONFIG_TOUCH_BUFSIZE);
+		if (!priv->tp_buf.buffer) {
+			return -ENOMEM;
```
Comment (sunghan-chang) [inline os/drivers/input/touchscreen.c:142]: Need give sem for line 134

File: os/drivers/input/ist415_misc.c [hunk 18]
```diff
@@ -0,0 +1,764 @@
...
+{
+	uint32_t *bufslf;
+	uint32_t *bufmtl;
+	int tx, rx;
+	uint32_t cdc, baseline, diff;
+	int ret = 0;
+
+	bufslf = (uint32_t *)kmm_malloc((int)dev->slf_node_len * sizeof(uint32_t));
+	bufmtl = (uint32_t *)kmm_malloc((int)dev->mtl_node_len * sizeof(uint32_t));
+
```
Comment (anjana348) [inline os/drivers/input/ist415_misc.c:416]: malloc failure not checked

File: os/drivers/input/ist415_misc.c [hunk 19]
```diff
@@ -0,0 +1,764 @@
...
+{
+	uint32_t *bufslf;
+	uint32_t *bufmtl;
+	int tx, rx;
+	uint32_t cpc;
+	int ret = 0;
+
+	bufslf = (uint32_t *)kmm_malloc(dev->slf_node_len * sizeof(uint32_t));
+	bufmtl = (uint32_t *)kmm_malloc(dev->mtl_node_len * sizeof(uint32_t));
+
```
Comment (anjana348) [inline os/drivers/input/ist415_misc.c:631]: malloc failure not checked

File: os/drivers/input/touchscreen.c [hunk 20]
```diff
@@ -88,7 +87,10 @@ static const struct file_operations g_touchdev_fileops = {
 	touch_write,	/* write */
 	NULL,			/* seek */
 	touch_ioctl,	/*ioctl */
+#if !defined(CONFIG_DISABLE_POLL)
 	touch_poll,
+#endif
+	NULL
```
Comment (sunghan-chang) [inline os/drivers/input/touchscreen.c:93]: Let's leave a comment for each like other lines.

touch_poll,  /* poll */

NULL   /* unlink */

File: os/drivers/input/ist415_misc.c [hunk 21]
```diff
@@ -0,0 +1,764 @@
...
+	int16_t *buf16;
+	int ret = 0;
+
+	ist415vdbg("*** SelfTest Start ***\n");
+
+	ist415_disable(dev);
+	ist415_sensor(dev, false);
+
+	buf16 = (int16_t *)kmm_malloc(dev->mtl_node_len * sizeof(int16_t));
+
```
Comment (anjana348) [inline os/drivers/input/ist415_misc.c:729]: malloc failure not checked

**Review by sunghan-chang — APPROVED** [review]

**Pull request merged**